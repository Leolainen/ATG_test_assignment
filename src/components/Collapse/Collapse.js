import React, { useState, useRef, useEffect } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import styles from "./styles.module.css";

const Collapse = (props) => {
  const {
    className,
    children,
    component: Component = "div",
    label,
    ...other
  } = props;
  const [collapsed, setCollapsed] = useState(true);
  const [wrapperHeight, setWrapperHeight] = useState(0);
  const wrapperRef = useRef(null);

  useEffect(() => {
    const currentHeight = wrapperRef.current
      ? wrapperRef.current.clientHeight
      : 0;

    setWrapperHeight(collapsed ? 0 : currentHeight);
  }, [collapsed]);

  const handleClick = () => {
    setCollapsed(!collapsed);
  };

  return (
    <Component
      className={classnames(
        styles.root,
        {
          [styles.expanded]: !collapsed,
        },
        className
      )}
      data-testid="collapse"
      {...other}
    >
      <div
        className={classnames(styles.header, {
          [styles.expanded]: !collapsed,
        })}
        data-testid="collapse-header"
        onClick={handleClick}
      >
        <h6>{label}</h6>

        <span
          className={classnames(styles.arrow, {
            [styles.rotate]: !collapsed,
          })}
        >
          â†“
        </span>
      </div>

      <div className={styles.wrapper} style={{ height: wrapperHeight }}>
        <div className={styles.innerWrapper} ref={wrapperRef}>
          {children}
        </div>
      </div>
    </Component>
  );
};

Collapse.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  label: PropTypes.string,
};

Collapse.uiName = "Collapse";

export default Collapse;
