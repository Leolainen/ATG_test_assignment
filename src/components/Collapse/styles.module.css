.root {
  padding: 0;
  margin: 0;
  background: white;
  overflow: hidden;
  border-radius: 3px;
  /* border: 1px solid var(--harmony-6); */
  box-shadow: 0 0.1px 0.5px rgba(0, 0, 0, 0.012),
    0 0.4px 1.6px rgba(0, 0, 0, 0.018), 0 2px 7px rgba(0, 0, 0, 0.03);

  transition: box-shadow 255ms ease-out;
}

.wrapper {
  transition: all 255ms ease-out;
}

.header {
  cursor: pointer;
  display: flex;
  align-items: center;
  padding: 8px;
  transition: all 125ms ease-out;
}

.header.expanded {
  background-color: var(--harmony-6);
}

.root.expanded {
  box-shadow: 0 0.3px 0.6px rgba(0, 0, 0, 0.036),
    0 0.9px 2px rgba(0, 0, 0, 0.054), 0 4px 9px rgba(0, 0, 0, 0.09);
}

.header > h6 {
  margin: 0;
  flex-grow: 1;
}

.header:hover {
  background-color: var(--harmony-6);
}

.arrow {
  transition: all 125ms ease-out;
  margin-right: 16px;
  transform-origin: center;
}

.rotate {
  transform: rotateZ(180deg);
}
