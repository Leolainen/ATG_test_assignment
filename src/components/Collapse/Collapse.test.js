import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import Collapse from "./Collapse";

describe("<Collapse />", () => {
  afterEach(cleanup);

  it("should render label text", () => {
    const { getByText } = render(<Collapse label="label">collapse</Collapse>);
    const label = getByText("label");

    expect(label).toBeInTheDocument();
  });

  describe("Snapshots", () => {
    it("should render as collapsed", () => {
      const { asFragment } = render(<Collapse>collapse</Collapse>);

      expect(asFragment()).toMatchSnapshot();
    });

    it("should render as expanded", () => {
      const { asFragment, getByTestId } = render(<Collapse>collapse</Collapse>);
      fireEvent.click(getByTestId("collapse-header"));

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
