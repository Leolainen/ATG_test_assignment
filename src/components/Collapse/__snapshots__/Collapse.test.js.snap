// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse /> Snapshots should render as collapsed 1`] = `
<DocumentFragment>
  <div
    class="root"
    data-testid="collapse"
  >
    <div
      class="header"
      data-testid="collapse-header"
    >
      <h6 />
      <span
        class="arrow"
      >
        ↓
      </span>
    </div>
    <div
      class="wrapper"
      style="height: 0px;"
    >
      <div
        class="innerWrapper"
      >
        collapse
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Collapse /> Snapshots should render as expanded 1`] = `
<DocumentFragment>
  <div
    class="root expanded"
    data-testid="collapse"
  >
    <div
      class="header expanded"
      data-testid="collapse-header"
    >
      <h6 />
      <span
        class="arrow rotate"
      >
        ↓
      </span>
    </div>
    <div
      class="wrapper"
      style="height: 0px;"
    >
      <div
        class="innerWrapper"
      >
        collapse
      </div>
    </div>
  </div>
</DocumentFragment>
`;
